### ISSUE TITLE
MacOS Failed to apply some or all settings

### ISSUE THREAD

**Issue by DonavanMartin** [de-vri-es/serial2-rs#44]
https://github.com/de-vri-es/serial2-rs/blob/4b9413dee17fd98a171d2969a7b7b32359b56d13/src/sys/unix/mod.rs#L163

I'm unable to connect port with MacOS: "/dev/tty.usbserial-DK0I847P": "ValidationError: StdError(Custom { kind: Other, error: \"failed to apply some or all settings\" })"

I think their's something misconfigured with baud rate.

Windows and Debian are working properly.

**Comment by bailout00** [comment]
Same issue here, some notes here:

https://github.com/de-vri-es/serial2-tokio-rs/issues/9

**Comment by bailout00** [comment]
A bit more info, definitely something around changing the baud rate. Changing any other settings seems to work normally, but changing the baud rate breaks it. 

Seems like it might be related to this possibly? 

https://github.com/de-vri-es/serial2-rs/issues/41

```
Serial port opened, settings: Settings { baud_rate: Ok(9600), char_size: Ok(Bits8), stop_bits: Ok(One), parity: Ok(None), flow_control: Ok(None) }
Error setting serial port settings: failed to apply some or all settings
Serial settings after set: Settings { baud_rate: Ok(115200), char_size: Ok(Bits8), stop_bits: Ok(One), parity: Ok(None), flow_control: Ok(None) }
```


```
match  SerialPort::open("/dev/tty.usbserial-16548", KeepSettings) {
        Ok(mut p) => {
            let mut settings = p.get_configuration().unwrap();
            debug!("Serial port opened, settings: {:?}", settings);
            settings.set_baud_rate(115200);
            let set_result = p.set_configuration(&settings);
            match set_result {
                Ok(_) => {
                    debug!("Serial port settings set: {:?}", settings);
                }
                Err(e) => {
                    println!("Error setting serial port settings: {}", e);
                    let mut settings = p.get_configuration().unwrap();
                    debug!("Serial settings after set: {:?}", settings);
                    
                }
            }
        }
        Err(e) => {
            println!("Error opening serial port {} {}", "/dev/tty.usbserial-14430", e);
        }
    }
```

Setting it back to 9600 DOES work but so far any other number fails

**Comment by de-vri-es** [comment]
Argh, I think it must be related to PR #39 (to fix issue #38).

Initial tests seemed to show that the custom speeds set that way are reported back in `tcgetattr`, but maybe this is not the case, and we read 9600 instead of the custom baud rate.

Sadly, if this is true, then there seems to be no way to retrieve the actual baud rate on OS X after we use the `IOCTL_IOSSIOSPEED` ioctl. And that means that using it will break `get_configuration()`.

To verify this, you can create the serial port with baud rate 9600, and then change the configuration afterwards. While it will report failure, it might actually have set the baud rate, so you might still be able to use the serial port.

If that works, could you see what `get_configuration()` reports and what `stty -F /dev/your-tty` says, *while you have your programming running with the serial port opened*?

**Comment by bailout00** [comment]
You're correct
```
$ stty -f /dev/tty.usbserial-16548
speed 115200 baud;
lflags: -icanon -isig -iexten -echo
iflags: -icrnl -ixon -ixany -imaxbel -brkint
oflags: -opost -onlcr -oxtabs
cflags: cs8 -parenb
```
And also get_configuration does show the correct speed even after the error message


Yea for now I've ignored the error on macos, the target is linux and so far its fine there so not a huge deal but should maybe be addressed in the lib somehow.

**Comment by de-vri-es** [comment]
I agree. I wonder is how stty manages to get the correct speed. Is it possible to use something like `strace` (but mac OS equivalent) to see what syscalls `stty` is performing?

**Comment by de-vri-es** [comment]
Could one of you test #45 on an Apple machine? Sadly I don't have access to one :(

**Comment by bailout00** [comment]
That seems good to me 👍

**Comment by de-vri-es** [comment]
Did you check if it works as expected? I'm still nervous the speed might not be reported back correctly with `tcgetattr`.

**Comment by bailout00** [comment]
Well, I actually can't use it at all on macos. Not sure why but it blocks forever when trying to write a single byte to any usb adapter so not sure I'd be a great test. But it did succeed and stty says the speed changed.

**Comment by de-vri-es** [comment]
Does it at-least not report an error on opening the serial port or applying the baud rate anymore?

**Comment by bailout00** [comment]
Correct, no longer throws an error

**Comment by de-vri-es** [comment]
Thanks for testing it! And thanks @DonavanMartin and @bailout00 for reporting the issue and providing information!

Released as v0.2.28. I will also bump `serial2-tokio`. :rocket:
